### DISCUSSION TITLE
Questions around <Threshold> tag for db linter

### DISCUSSION THREAD

**Discussion by bheemreddy181-zz** [schemacrawler/SchemaCrawler#486] [Q&A] [answered]
Schema Crawler Config Used 

```
<schemacrawler-linter-configs>
  <linter id="io.github.mbarre.schemacrawler.tool.linter.LinterTimeStampWithOutTimeZoneColumn">
    <run>true</run>
    <severity>critical</severity>
    <threshold>807</threshold>
  </linter>
  <linter id="schemacrawler.tools.linter.LinterTableWithNoPrimaryKey">
    <run>true</run>
    <severity>critical</severity>
    <threshold>175</threshold>
  </linter>
</schemacrawler-linter-configs>
```
With the above threshold in place, I assume I should see the linter to show all failures but the overall job to be passed with the threshold, I feel the threshold tag works the same as exclude-columns 

Linter command used : 

```
schemacrawler -server=postgresql -database=cnuapp_int -user=postgres -password='' -linterconfigs=schemacrawler-linter-configs.xml -c=lint -lintdispatch=terminate_system -runalllinters=false
```

CI run looks like below 

<img width="1636" alt="Screen Shot 2021-07-08 at 11 29 00 AM" src="https://user-images.githubusercontent.com/3464623/124958415-befec200-dfdf-11eb-8261-419efd2895d8.png">

Expecting to see the CI run more like below but with an exit code 0

<img width="1444" alt="Screen Shot 2021-07-08 at 11 30 22 AM" src="https://user-images.githubusercontent.com/3464623/124958612-f53c4180-dfdf-11eb-9ee1-a3393997d8ea.png">

Would like to hear your thoughts @sualeh

**Comment by sualeh** [comment]
@bheemreddy181 - I want to see if the threshold is working. Please make both thresholds 10, and try again. Then make them both 3000 and try again. Please let me know what happens. 

Also, I will only be supporting SchemaCrawler 16.15.2. Please use the latest version of SchemaCrawler, so I can help you debug your issues. (Note that the command-line switches have changed from what you are using.)

**Reply by bheemreddy181-zz** [reply]
We need an upgrade here https://github.com/adriens/schemacrawler-deb/issues/193

**Reply by sualeh** [reply]
@bheemreddy181 - try this: https://github.com/schemacrawler/SchemaCrawler-Installers/releases/tag/v16.15.2

**Reply by bheemreddy181-zz** [reply]
But we need additional lints package as we use lints from there

**Reply by sualeh** [reply]
I see. The we will have to wait for @adriens to make it available for https://github.com/adriens/schemacrawler-deb/issues/193. 

I am curious - did you try out the regular SchemaCrawler installer from https://github.com/schemacrawler/SchemaCrawler-Installers/releases/tag/v16.15.2? Did it work?

**Reply by bheemreddy181-zz** [reply]
Ah no , we have to make a good amount of changes to switch from additional lints to just schema crawler - I will probably wait for @adriens to make that package available

**Reply by sualeh** [reply]
@bheemreddy181 - In any case, let us see if the threshold is working. Please make both thresholds 10, and try again. Then make them both 3000 and try again. Please let me know what happens.

**Reply by bheemreddy181-zz** [reply]
@sualeh I did try the way you mentioned above here are the results 

when the  `<threshold>10<threshold>` it still fails 

<img width="1474" alt="Screen Shot 2021-07-12 at 4 29 37 PM" src="https://user-images.githubusercontent.com/3464623/125358784-fc8c8380-e32e-11eb-86de-c09fcdbe4a9d.png">

when the  `<threshold>3000<threshold>` it's a `Success`

<img width="1398" alt="Screen Shot 2021-07-12 at 4 33 04 PM" src="https://user-images.githubusercontent.com/3464623/125358841-129a4400-e32f-11eb-9873-42fd0ac6d44e.png">

**Reply by sualeh** [reply]
Success meaning error code of 0?

**Reply by bheemreddy181-zz** [reply]
Yes @sualeh

**Reply by bheemreddy181-zz** [reply]
What is so significant about these two numbers? ( i.e 10 and 50 )

**Reply by sualeh** [reply]
Nothing really. Trying to understand what is happening, and how I can reproduce it.

**Comment by sualeh** [comment, marked as answer]
@bheemreddy181 - I tried, but I could not reproduce your issue. I think what you will have to do is use the latest version of SchemaCrawler, and try to reproduce. If the issue persists, we may have to do a screen share to see what is going on.

**Reply by bheemreddy181-zz** [reply]
Can you specify a set of linter's which are Postgres specific so I can just use the schema crawler and check if that works as expected @sualeh

**Reply by sualeh** [reply]
Any of the built-in ones will work with PostgreSQL.

**Reply by sualeh** [reply]
Use the latest SchemaCrawler. First just run the lint command without any arguments. Find the linter that identifies some lints. Create a YAML linter config file for that linter, and set a threshold.

**Reply by bheemreddy181-zz** [reply]
makes sense

**Reply by bheemreddy181-zz** [reply]
@sualeh am i missing something here 

```
./schemacrawler.sh -server=postgresql -database=pgs_development -user=postgres -password=''  -c=lint -lintdispatch=terminate_system --info-level=standard --log-level=CONFIG
```
this command yields

```
Jul 13, 2021 9:18:59 PM schemacrawler.tools.commandline.utility.CommandLineLoggingUtility logSafeArguments
INFO: Command line: 
-server=postgresql
-database=pgs_development
-user=postgres
-password=*****
-c=lint
-lintdispatch=terminate_system
--info-level=standard
--log-level=CONFIG
Jul 13, 2021 9:18:59 PM schemacrawler.tools.commandline.utility.CommandLineLoggingUtility logFullStackTrace
SEVERE: No database connection options provided
picocli.CommandLine$ParameterException: No database connection options provided
	at schemacrawler.tools.commandline.command.ConnectCommand.getDatabaseConnectionOptions(ConnectCommand.java:97)
	at schemacrawler.tools.commandline.command.ConnectCommand.run(ConnectCommand.java:115)
	at schemacrawler.tools.commandline.SchemaCrawlerCommandLine.executeCommandLine(SchemaCrawlerCommandLine.java:99)
	at schemacrawler.tools.commandline.SchemaCrawlerCommandLine.execute(SchemaCrawlerCommandLine.java:67)
	at schemacrawler.Main.main(Main.java:73)
```

**Reply by bheemreddy181-zz** [reply]
do i need a database url to connect to a database ? @sualeh

**Reply by sualeh** [reply]
@bheemreddy181 - 
1. The command-line arguments have changed in the latest version. They are now POSIX compliant. Please use the new command-line arguments.
2. The linter config file is now in YAML format. Please use the new YAML format.

**Reply by sualeh** [reply]
```sh
./schemacrawler.sh \
--server=postgresql \
--database=pgs_development \
--user=postgres --password=***** \
--info-level=standard \
--command=lint \
--lint-dispatch=terminate_system \
--run-all-linters=false \
--linter-configs schemacrawler-linter-configs.yaml
```

**Reply by sualeh** [reply]
Example YAML file:

```yaml
- id: schemacrawler.tools.linter.LinterTableWithNoRemarks
  run: true
  severity: critical
  threshold: 32
```

**Reply by bheemreddy181-zz** [reply]
Odd that it works on the newer version but doesn't understand if system exited with error code 0 or 1 , below are the details 

`<threshold>934<threshold>` for 935 failed lints 
<img width="915" alt="Screen Shot 2021-07-14 at 11 06 08 AM" src="https://user-images.githubusercontent.com/3464623/125655283-55a06ecf-1be6-46e4-85f7-848ef4061434.png">

`<threshold>935<threshold>` for 935 failed lints 
<img width="1057" alt="Screen Shot 2021-07-14 at 11 06 52 AM" src="https://user-images.githubusercontent.com/3464623/125655325-fd127562-1e62-4efb-b9d9-d2170488f385.png">

Can we add a success message at the end when the threshold and the number of failed lints matches

**Reply by bheemreddy181-zz** [reply]
@adriens a newer upgrade will help me testing additional lints as well - it would be great if you can dedicate some time.

**Reply by sualeh** [reply]
If you want to see the success or failure message, run it again with an extra `--log-level=INFO`

**Reply by bheemreddy181-zz** [reply]
Still not clear i feel with threshold-935

<img width="881" alt="Screen Shot 2021-07-14 at 11 30 01 AM" src="https://user-images.githubusercontent.com/3464623/125658516-6ca1ae45-14fa-49b1-81a4-8c57c3ce9475.png">

**Reply by sualeh** [reply]
Ok, I will make it clearer in the next release.

**Reply by sualeh** [reply]
Meanwhile, run the SchemaCrawler command, and then run `printf '%d\n' $?` to see the exit code.

**Reply by bheemreddy181-zz** [reply]
if the exit codes are correct with the above command - then both the thresholds are returning exit code `1` which means a failure - seems not working on the new version also

**Reply by bheemreddy181-zz** [reply]
Happy to do a screen share if needed

**Reply by sualeh** [reply]
@bheemreddy181 - please send me the exact command-line you used, and the exact contents of the linter configs file.

**Reply by bheemreddy181-zz** [reply]
I am running this on mac from the folder 

`/Users/user/Downloads/schemacrawler-16.15.2-distribution/_schemacrawler`

Command-line used 

```
 ./schemacrawler.sh --server=postgresql --database=pgs_development --password='' --info-level=standard --command=lint --lint-dispatch=terminate_system --run-all-linters=false --linter-configs schemacrawler-linter-configs.yaml --log-level=INFO
```
Config used 

```
- id: schemacrawler.tools.linter.LinterTableWithNoRemarks
  run: true
  severity: critical
  threshold: 934
```
or 
```
- id: schemacrawler.tools.linter.LinterTableWithNoRemarks
  run: true
  severity: critical
  threshold: 935
```

**Reply by sualeh** [reply]
And in both cases, you got identical results? Are you sure the linter configs file is getting picked up? You will have to go through the logs to find out. Can you run with `--log-level=ALL` and send me the logs (zipped) by email?

**Reply by bheemreddy181-zz** [reply]
Ah sorry - you are correct - exit codes were different - for some reason, I read them wrong last time - i could see exit code 0 for 935 vs exit code 1 for 934. Working as expected and config is getting read correctly.

**Reply by sualeh** [reply]
Perfect. Issue resolved.

**Reply by bheemreddy181-zz** [reply]
will wait for Additional lints to be updated and test there too.

**Reply by sualeh** [reply]
The way to distinguish them is you will get a message saying "Too many schema lints were found" in the logs. This message will be logged at a different severity level based on the type of dispatch.

**Reply by sualeh** [reply]
Ok. You can now mark this discussion as answered, so we can close it out.
